agents/timestream_aggregator/timestream_aggregator.py

```diff
@@ -24,28 +30,319 @@ def _create_file_path(start_time, data_dir):
+
+    Arguments
+    ---------
+    dirname : str
+        dirname for output, likely from _create_dirname
+    basename : str
+        basename of file without extension or suffix
+    suffix : str
+        Suffix to append to filename (include your own '_')
+    extension : str
+        file extension to put on file (without '.')
+
+    Returns
+    -------
+    str
+        Full filepath of file for output
+
+    """
+    filename = "{}{}.{}".format(basename, suffix, extension)
+    filepath = os.path.join(dirname, filename)
 
     return filepath
 
 
+class FrameRecorder:
+    """Frame recording object. Keeps track of the reader/writer and associated
+    information, such as the frames being written to file and the location of
+    the output files.
+
+    All of the methods were separated from the loop that ran in the
+    TimestreamAggregator OCS Agent. They're meant to be looped over still and
+    for that we provide the run() method.
+
+    Parameters
+    ----------
+    file_duration : int
+        seconds per file
+    tcp_addr : str
+        tcp connection address (i.e. "tcp://127.0.0.1:4536")
+    data_dir : str
+        Location to write data files to. Subdirectories will be created within
+        this directory roughly corresponding to one day.
+    log : txaio.tx.Logger
+        txaio logger object, created by the OCSAgent.
+
+    Attributes
+    ----------
+    time_per_file : int
+        Amount of time in seconds to have in each file. Files will be rotated
+        after this many seconds.
+    address : str
+        Address of the host sending the data via the G3NetworkSender.
+    data_dir : str
+        Location to write data files to. Subdirectories will be created within
+        this directory roughly corresponding to one day.
+    log : txaio.tx.Logger
+        txaio logger object, created by the OCSAgent.
+    reader : spt3g.core.G3Reader
+        G3Reader object to read the frames from the G3NetworkSender.
+    writer : spt3g.core.G3Writer
+        G3Writer for writing the frames to disk.
+    frames : list
+        List of frames that have been read from the network. Gets cleared after
+        writing to file.
+    last_meta
+        The last meta data Frame passed to the G3Writer. The last meta data
+        frame gets written to the top of each file.
+    last_frame_write_time : float
+        The time at which we wrote the last frame. Used to track when files
+        should be rotated between acquisitions.
+    filename_suffix : int
+        For enumerating files within an acquisition. We want to split files in
+        10 minute intervals, but the names should remain the same with a
+        "_0001" style suffix, which we track with this attribute.
+    dirname : str
+        The directory path currently in use.
+    basename : str
+        The basename of the file currently being used. The actual file will
+        also contain the tracked suffix.
+
```

Comment:
Might be worth adding start_time here, and pointing out that it's the timestamp of the current file, not the whole fileset (i.e. it will only agree with basename when filename_suffix=0).

Reply:
Thanks for pointing that, I missed documenting that attribute. Added now.